### ISSUE TITLE
Yarn workspaces

### ISSUE THREAD

**Issue by edorivai** [withspectrum/spectrum#2952]
Yarn [workspaces](https://yarnpkg.com/lang/en/docs/workspaces/) should simplify the dependency management within the monorepo.

I'm going to give this thing a naive try, meaning I'll enable workspaces, and see if I can get all of the dependencies to install with a single `yarn` command.

If somebody could chime in on the implications for CI and deployment, that would be great :+1:

**Comment by mxstbr** [comment]
This is the main blocker: https://github.com/zeit/now-cli/issues/938 We can't deploy when moved to yarn workspaces

**Comment by mxstbr** [comment]
> I'd have every package.json be atomic, and declare its own dependencies.

I mean, that makes sense to me. If that makes it work then let's do it? Also I think #2793 added something related to workspaces? /cc @fubhy

**Comment by edorivai** [comment]
Actually, I see that you're copying the root `package.json` and `now.json` to each of the services:

```
"postbuild:pluto": "cp pluto/package.json build-pluto/package.json && cp now.json build-pluto/now.json",
```
Do you think it work with Now if we'd do the same with the root `yarn.lock`?

**Comment by mxstbr** [comment]
If we did that then every service would install all the dependencies of all the services, right? Since the root `yarn.lock` will contain all dependencies?

**Comment by edorivai** [comment]
Well, not according to https://twitter.com/bestander_nz/status/960920084410654720

> You should definitely be able to reuse the root yarn.lock in a workspace in isolation, the extra resolutions will be ignored

**Comment by edorivai** [comment]
I think yarn will respect the dependencies in `package.json`, and just uses the `yarn.lock` as a reference to look up the "preferred" version for the packages defined in `package.json`.

**Comment by edorivai** [comment]
I just opened a PR: #2953, maybe best to continue discussion there?

**Comment by fubhy** [comment]
I am also adding workspaces in https://github.com/withspectrum/spectrum/pull/2793. Can we wait with this PR here? I've already been working on the #2793 for a considerable amount of time.

**Comment by edorivai** [comment]
@fubhy Yeah, the code in this PR is relatively trivial (at least until now). If you manage to get it all up and running in your PR, including the deployment to Now, this PR can de discarded :smile:

Do you have an idea of how long you need to get #2793 mergable? If you need a considerable amount of time, it might be worth it to tackle the workspaces separately here. I'm guessing it should be relatively easy to merge the changes from this PR into yours. What do you think?

**Comment by fubhy** [comment]
The deployment to Now is orthogonal to the PR because the way we are going to deploy to Now with my PR is by simply submitting a Docker image. So it's already solved imho.

As far as I am concerned, my PR is already mergeable. The only thing that keeps us from doing that for now is that Max reported issues with the bootstrap phase of the images.

**Comment by edorivai** [comment]
Ah, that's good news, I didn't consider the impact of docker containers on the Now deployment. That will simplify things :). I'll watch the progress on #2793 :eyes:

**Comment by sibelius** [comment]
maybe nohoist could help https://yarnpkg.com/blog/2018/02/15/nohoist/

also using lerna?

**Comment by kbrandwijk** [comment]
So, I recently jumped on the Spectrum train trying to get everything running locally and self-hosting. One of the first things I wondered where regarding Docker and workspaces. Then, I ran across this issue, that is marked as Blocked Upstream, but workspaces are already supported by Now 2.0. Also, work stranded here because of the Docker PR, but at the same time, the Docker PR was closed, because it was decided not to use Docker.

Does that mean that, with Docker and with Now 2.0 support for workspaces, this could be an option again?
